### PR TITLE
Make `Transformation` and `NonTransformation` subclass `TransformationBase`

### PR DESCRIPTION
**Pull request by dotsdl**
In some cases it can be awkward for `NonTransformation` to be a subclass of `Transformation`, such as in `alchemiscale`, for cases where `NonTransformation` should be handled very differently.

Switching to a shared, abstract base class for `Transformation` and `NonTransformation` simplifies this.

### CODE REVIEW

**Comment by dotsdl** [comment]
I can do that, no problem @IAlibay! I also need to fix the broken tests yet as well.

**Comment by IAlibay** [comment]
@dotsdl possibly a crazy question - does this technically change the gufe keys for transformation objects? I can't remember if it captures the parent class signatures

(partly asking because of the test failures)

**Comment by IAlibay** [comment]
The follow up question I have is if we need a 1.x gufe policy on key stability?

**Comment by dotsdl** [comment]
Will pursue a review on this from Alyssa when she starts.

**Review by IAlibay — COMMENTED** [review]
This looks good to me, couple of comments though.

File: gufe/transformations/transformation.py [hunk 1]
```diff
@@ -13,7 +14,102 @@
 from ..mapping import ComponentMapping
 
 
-class Transformation(GufeTokenizable):
+class TransformationBase(GufeTokenizable):
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply gufe/transformations/transformation.py:17]: Done!

File: gufe/transformations/transformation.py [hunk 2]
```diff
@@ -211,8 +236,13 @@ def __init__(
     ):
 
         self._system = system
-        self._name = name
         self._protocol = protocol
+        self._name = name
+
+    def __repr__(self):
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply gufe/transformations/transformation.py:267]: Added!

**Comment by dotsdl** [comment]
@IAlibay can I get another review from you? I also used this as an opportunity to clean up some docstrings, and adjust the reprs for `Transformation` and `NonTransformation` to include the `name` attribute.

**Review by IAlibay — APPROVED** [review]
lgtm thanks!

**Review by atravitz — APPROVED** [review]
This refactor is a big improvement imo, just a couple docstrings and clarifying questions.

File: gufe/transformations/transformation.py [hunk 3]
```diff
@@ -178,81 +128,188 @@ def load(cls, file):
...
 
-# we subclass `Transformation` here for typing simplicity
-class NonTransformation(Transformation):
-    """A non-alchemical edge of an alchemical network.
+class Transformation(TransformationBase):
+    _stateA: ChemicalSystem
+    _stateB: ChemicalSystem
+    _name: Optional[str]
+    _mapping: Optional[Union[ComponentMapping, list[ComponentMapping]]]
+    _protocol: Protocol
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply gufe/transformations/transformation.py:144]: Done!

**Comment by dotsdl** [comment]
pre-commit.ci autofix

**Comment by dotsdl** [comment]
Thanks @IAlibay and @atravitz! Merged!

**Pull request merged**